File: clojure/game-of-life/src/game_of_life.clj
(ns game-of-life)

(defn- get-cell [[row col] cells]
  "Returns the given cell value, or nil if out of bounds."
  (when
    (and (<= 0 row (dec (count cells))) (<= 0 col (dec (count (first cells)))))
    (get-in cells [row col])))

(defn- neighbour-count [[row col] cells]
  "Returns the count of 1s neigbouring the given cell."
  (reduce + 
    (keep #(get-cell % cells)
          [[(dec row) (dec col)] [(dec row) col] [(dec row) (inc col)]
           [row       (dec col)]                 [row       (inc col)]
           [(inc row) (dec col)] [(inc row) col] [(inc row) (inc col)]])))

(defn- tick-cell [[row col] cells]
  "Returns the next generation of the given cell."
  (let [cell-value (get-cell [row col] cells)
        cell-neighbours (neighbour-count [row col] cells)]
    (cond
      (and (= 1 cell-value) (<= 2 cell-neighbours 3)) 1
      (and (= 0 cell-value) (= 3 cell-neighbours)) 1
      :else 0)))

(defn tick
  "Returns the next generation of the cells."
  [cells]
  (mapv (fn [row]
          (mapv (fn [col]
                  (tick-cell [row col] cells))
                (range (count (first cells)))))
        (range (count cells))))

File: python/game-of-life/game_of_life.py
"""Implement Conway's Game of Life"""

def neighbours(target_row, target_column, matrix):
    """Returns the number of live neighbours of a cell"""
    height, width = len(matrix), len(matrix[0])
    return sum(
        matrix[row][column]
        for row in range(max(0, target_row - 1), min(height, target_row + 2))
        for column in range(max(0, target_column - 1), min(width, target_column + 2))
        if (row, column) != (target_row, target_column)
    )


def tick_cell(cell, neighbour_count):
    """Returns the new state of a single cell, given its neighbour count"""
    return int(neighbour_count == 3 or (cell == 1 and neighbour_count == 2))


def tick(matrix):
    """Returns the new state of the entire matrix after one tick"""
    return [
        [
            tick_cell(matrix[row][column], neighbours(row, column, matrix))
            for column in range(len(matrix[row]))
        ]
        for row in range(len(matrix))
    ]
    